docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,72 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
 **s2n_config_add_ticket_crypto_key** adds session ticket key on the server side. It would be ideal to add new keys after every (encrypt_decrypt_key_lifetime_in_nanos/2) nanos because
 this will allow for gradual and linear transition of a key from encrypt-decrypt state to decrypt-only state.
 
+### Asynchronous private key operations related calls
+
+When s2n is used in non-blocking mode, this set of functions allows user
+to move execution of CPU-heavy private key operations out of the main
+event loop, preventing **s2n_negotiate** blocking the loop for a few
+milliseconds each time the private key operation needs to be performed.
+
+To enable asynchronous private key operations user needs to provide a
+callback function **s2n_async_pkey_fn** to
+**s2n_config_set_async_pkey_callback** call. This function will be
+executed each time operation on private key needs to be performed with
+**op** object representing this operation. From this function the user
+can spawn the thread with **op** object to perform through
+**s2n_async_pkey_op_perform** call and immediately return from the
+function without waiting for thread to complete. The **s2n_negotiate**
+will fail with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status**
+**S2N_BLOCKED_ON_APPLICATION_INPUT**, and will keep giving the same
+error until the **op** is performed on the spawned thread and applied to
+connection. Once the spawned thread finishes performing the **op**,
+the **s2n_async_pkey_op_apply** needs to be called in the main
+event loop thread to apply the result to connection. After that the user
+needs to call **s2n_negotiate** to proceed with handshake.
+
+```c
+typedef int (*s2n_async_pkey_fn)(struct s2n_connection *conn, struct s2n_async_pkey_op *op);
+extern int s2n_config_set_async_pkey_callback(struct s2n_config *config, s2n_async_pkey_fn fn);
+extern int s2n_async_pkey_op_perform(struct s2n_async_pkey_op *op, s2n_cert_private_key *key);
+extern int s2n_async_pkey_op_apply(struct s2n_async_pkey_op *op, struct s2n_connection *conn);
+extern int s2n_async_pkey_op_free(struct s2n_async_pkey_op *op);
+```
+
+- **op** is an opaque object representing private key operation which
+needs to be performed.
+- **key** is a private key used for operation.
+
+**s2n_async_pkey_fn** is invoked every time some action involving
+private key is required. The **conn** provides a pointer to connection
+which triggered this action. The **op** is an operation which needs to be
+performed through **s2n_async_pkey_op_perform** call and applied to
+**conn** through **s2n_async_pkey_op_apply** call after the
+callback was complete. Once performed and applied the **op** should be
+freed by the caller through **s2n_async_pkey_op_free** call.
```

Comment:
One more thing worth documenting is whether users are allowed to call the `s2n_async_pkey_op_perform` and then `s2n_async_pkey_op_apply` synchronously from inside the callback. I guess not, since it might lead to weird reentrancy issues.

Reply:
Yup, that's not allowed. The benefit from allowing it didn't justify the number of scenarios which need to be taken care of to support that. Will add this to the doc.